### PR TITLE
Annotation: always use FancyBboxPatch instead of bbox_artist

### LINKED ISSUE #4139: Annotation text bbox calculated incorrectly in matplotlib>=1.4.3?
Prior to matplotlib 1.4.3, text bounding boxes created with `annotate` did not include the connection arrow.  Starting with 1.4.3 and continuing up to the current HEAD, the bbox includes the connection arrow.  

As an example:

```
import matplotlib.pyplot as plt
import matplotlib

fig, ax = plt.subplots()

ax.annotate('Matplotlib {}'.format(matplotlib.__version__),
            xy=(0.1, 0.1), xytext=(50, 50), size=20,
            xycoords='data', textcoords='offset points',
            bbox=dict(alpha=0.5, fc='yellow'),
            arrowprops=dict(arrowstyle='->'))

plt.show()
```

mpl1 4 2

mpl1 4 3

mpl_github

This may be related to https://github.com/matplotlib/matplotlib/issues/4012?

### LINKED ISSUE #4140: Annotation bbox clipping incorrect for `boxstyle='round'` Matplotlib >= 1.4.0
Annotations using a bbox boxstyle of "round" have different (incorrect?) clipping behavior compared to a square boxstyle.

As an example, note that with matplotlib 1.4.0, the connector arrow extends inside the round box, while it is clipped for the square boxstyle.  With earlier versions, both have the same clipping behavior.

```
import matplotlib.pyplot as plt
import matplotlib

def main():
    fig, ax = plt.subplots()
    ax.set(title='Matplotlib {}'.format(matplotlib.__version__))
    annotate(ax, 'Round')
    annotate(ax, 'Square')
    plt.show()

def annotate(ax, text):
    bbox = dict(fc='none')
    x, y = 0.1, 0.1

    if text == 'Round':
        bbox['boxstyle'] = 'round'
        y = 0.5

    ax.annotate(text, xy=(x, y), xytext=(50, 50), size=20,
                xycoords='data', textcoords='offset points',
                bbox=bbox,  arrowprops=dict(arrowstyle='->'))

main()
```

clip_mpl1 3

clip_mpl1 4 0

The problem persists up to and including the current HEAD, but newer versions also have the issue described in https://github.com/matplotlib/matplotlib/issues/4139  The two issues don't appear to be directly related.

### PR DESCRIPTION
**Pull request by efiring**
This fixes #4139, as an alternative to #4145, by eliminating the use of the bbox_artist debugging function.  With a second commit, it also fixes #4140.

### CODE REVIEW

**Comment by efiring** [comment]
I'm not entirely happy with the second commit because I still don't understand how, when, and where the transforms are evolving.  I determined that prior to this commit the `patchA.get_transform()` being used during clipping and shrinking of the arrow path was incorrect, but it was correct after everything was drawn.  I then guessed that drawing the text, along with patchA, would update patchA's transform, and that turned out to be the case.
There are still some problems.  First, switching from bbox_artist to a FancyBboxPatch makes a larger box. I suspect (or at least hope) this will be easy to fix.  Second, when a YAArrow is used instead of a FancyArrowPatch, it's end is located based on the text bbox, not the FancyBboxPatch boundary, so it tends to extend into the Patch.  This is not new.
I haven't looked in the tests directory, but I get the impression that we have quite a body of very complex and hard to understand text/annotation/patch code with very little in the way of automated tests.

**Comment by tacaswell** [comment]
I am regretting both my `Annotation` refactor from last cycle and merging the bounding box fix into 1.4.x as it seems to have opened a can of worms we were happily ignoring for years.

**Comment by efiring** [comment]
The Travis failure is because the pylab_examples/text_rotation.py is including a "pad" in a dictionary provided as the 4th arg to `text()`. The way this argument/kwarg is being handled, and described, is confusing, to say the least.  Try tracing it through pyplot to axes to Text.  Before my change it was being popped off of a copy of a dictionary, and I deleted all that.  Evidently a pad argument needs to go back in, presumably to control the padding of the 'square' FancyBboxPatch.

**Comment by efiring** [comment]
@tacaswell, worms probably would have started crawling out sooner or later; but my dislike for working in this part of the code, which I have mostly avoided, has been strongly reinforced.

**Comment by efiring** [comment]
Adjusting the size of the FancyBboxPatch appears to be rather difficult and intrusive.  I tried what seemed like a straightforward way to remove some of the extra space below the text--and I failed miserably.
A side effect of dropping the bbox_artist option is that the text box patch now rotates with the text, so pylab_examples/text_rotation.py yields a very different picture.  I think that the FBP behavior is what a user would want, so I don't think that losing the old behavior, in which the unrotated extent rectangle is used, will be missed by many people.

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -3897,7 +3897,7 @@ def __init__(self, posA=None, posB=None,
         """
         If *posA* and *posB* is given, a path connecting two point are
         created according to the connectionstyle. The path will be
-        clipped with *patchA* and *patchB* and further shirnked by
+        clipped with *patchA* and *patchB* and further shrinked by
```
Comment (tacaswell) [inline lib/matplotlib/patches.py]: shrunk?

File: lib/matplotlib/text.py [hunk 2]
```diff
@@ -823,10 +814,10 @@ def set_backgroundcolor(self, color):
 
         ACCEPTS: any matplotlib color
         """
-        if self._bbox is None:
-            self._bbox = dict(facecolor=color, edgecolor=color)
+        if self._bbox_patch is None:
+            self.set_bbox = dict(facecolor=color, edgecolor=color)
         else:
-            self._bbox.update(dict(facecolor=color))
+            self._bbox_patch.update(dict(facecolor=color))
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: I think this should be de-dented

File: lib/matplotlib/text.py [hunk 3]
```diff
@@ -823,10 +814,10 @@ def set_backgroundcolor(self, color):
 
         ACCEPTS: any matplotlib color
         """
-        if self._bbox is None:
-            self._bbox = dict(facecolor=color, edgecolor=color)
+        if self._bbox_patch is None:
+            self.set_bbox = dict(facecolor=color, edgecolor=color)
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: I think this should be a function call?

Comment (efiring) [inline lib/matplotlib/text.py]: Correct.

File: lib/matplotlib/text.py (hunk 2 above)
Comment (efiring) [inline lib/matplotlib/text.py]: Why?  As it is, the update of the existing _bbox_patch is needed only if one hasn't just created a new one with the desired color.

File: lib/matplotlib/text.py [hunk 4]
```diff
@@ -825,10 +816,10 @@ def set_backgroundcolor(self, color):
 
         ACCEPTS: any matplotlib color
         """
-        if self._bbox is None:
-            self._bbox = dict(facecolor=color, edgecolor=color)
+        if self._bbox_patch is None:
+            self.set_bbox(dict(facecolor=color, edgecolor=color))
         else:
-            self._bbox.update(dict(facecolor=color))
+            self._bbox_patch.update(dict(facecolor=color))
```
Comment (efiring) [inline lib/matplotlib/text.py]: @tacaswell, your comment on this line, and my response, were lost when I rebased.  You thought the line should be dedented, but I don't understand why.  The update is needed only if a new _bbox_patch has not just been created.

Comment (tacaswell) [inline lib/matplotlib/text.py]: I don't remember exactly what I was thinking when I left the last message.  I think what I was suggesting was to get rid of the `else` block and just always update the `_bbox_patch` (even if we just created it with the correct color).  It is a no-op and I think makes the code path a little easier to follow (which is a style call).

Avoiding `else` statements and having `if` blocks which just ensure assumptions are true is is a newish (and possibly not good) style-tick I have developed. 

I think my comment on this can be ignored.

**Comment by tacaswell** [comment]
I think this also closes #4012

**Comment by tacaswell** [comment]
Can you add an API changes note?

We probably want to back-port this to color-overhaul as I broke this by merging what I thought were bug-fixes into 1.4.3.

**Comment by efiring** [comment]
This is getting close to ready. @mdboom or @tacaswell, the remaining problem is that the arrow starts inside the patch.  It would be nice to be able to use the patch boundary to clip the arrow.  Is there an easy way to do this?

**Comment by efiring** [comment]
This is just too complicated.  I think something is going wrong in the chain of events leading to the clip method of the Base class inside the ConnectionStyle class which is called by FancyArrowPatch.  My guess is that some transform is being changed too early, or too late, so that the attempted clipping is working with a bbox_patch and an arrow_patch that are not being transformed the same way.  I can't spend any more time on it. @mdboom, will you be able to take a look?  Maybe at the same time you will be able to figure out a way to make all this less convoluted and hard to follow.  The capabilities provided by all the Fancy things are wonderful, but figuring out how it all works--or, at present, doesn't quite work--is extraordinarily difficult.  Maybe all it needs is some overview documentation or additional comments.

**Comment by efiring** [comment]
The Travis failure is because of a subtle change in arrow patch location and size.  We can either change the test images or change the patch location and size for these cases.  I know how to do either; this is not a problem.  In fact, I suspect the aesthetic optimum might be half way between the two cases; previously there was a little too much space below the text, because the full descent was being added.  I took that out to get a better match to what the bbox_artist was doing.

**Comment by tacaswell** [comment]
attn @leejjoon.  I think you wrote the original version of most of this code, any guidance on where it is going wrong?

**Comment by efiring** [comment]
I rebased this to master, but I'm still struggling with the problem that the YAArrow extends to the text bbox, which in general is inside the fancy patch.  I tried setting a direction-reversed fancy patch path as a clip region, hoping this would mean the *inside* of the polygon would be the blanked-out region, but that didn't work.  The solution might be to abandon the YAArrow and simulate it with a FancyArrowPatch. An advantage would be that this would mean that the arrow in Annotation would always be a FancyArrowPatch.  The trick would be to make the API and behavior similar enough to the original YAArrow version to keep people happy.

**Comment by tacaswell** [comment]
:+1: in principle to dropping YAArrow and just using FanceArrowPatch for
everything.  The variability of that code path bothers me.

**Comment by efiring** [comment]
I think this is getting close to being ready.  It won't pass the tests, but we have to figure out where this means the code needs adjustment, and where we need new test images.  There is also some more documentation and cleanup to be done.  The YAArrow replacement does not *exactly* reproduce the old behavior and API, but I think it will be close enough, and noticeably improved.  It removes the embarrassing aspect of annotation_demo.py in which the short arrow has an absurdly short head.

**Comment by efiring** [comment]
I think this is ready to go.  The failure on python 2.6 seems random and unrelated; I restarted it, and got a completely different failure than the time before.  I will try restarting it one more time.

File: lib/matplotlib/text.py [hunk 5]
```diff
@@ -2246,16 +2242,13 @@ def draw(self, renderer):
...
-
-        if self.arrow_patch is not None:
+        if self.arrow_patch is not None:   # FancyArrowPatch
             if self.arrow_patch.figure is None and self.figure is not None:
                 self.arrow_patch.figure = self.figure
             self.arrow_patch.draw(renderer)
 
+        # Draw text, including FancyBboxPatch, after FancyArrowPatch.
+        # Otherwise, a wedge arrowstyle can land partly on top of the Bbox.
         Text.draw(self, renderer)
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: use super here (you use it in `update`)?

Comment (efiring) [inline lib/matplotlib/text.py]: True, I did.  That is new code, and right next to another method in which super is used.  But the "Text.draw()" is not new, and it is one of several such lines in this module.  I'm inclined to think that "super-ifying" the module is out of scope for this PR.  I also note that not all parent class method calls in this module can be handled with super;  `Annotation.__init__` needs to call the `__init__` from each of its immediate parents.  The only way to avoid that would be to rewrite it so as to use composition rather than multiple inheritance.  In the long run that could bring a big gain in readability and maintainability, but given the present complexity of this entire family of code, it would be quite a job.

Comment (tacaswell) [inline lib/matplotlib/text.py]: Fair.
